src-self-hosted/translate_c.zig

```diff
@@ -924,6 +1053,327 @@ fn transImplicitCastExpr(
     }
 }
 
+fn toEnumZeroCmp(
+    rp: RestorePoint,
+    scope: *Scope,
+    expr: *ast.Node,
+    generate_enum_node: fn (RestorePoint, *const struct_ZigClangType, source_loc: ZigClangSourceLocation) TransError!*ast.Node,
+    enum_ty: *const struct_ZigClangType,
+    enum_source_loc: ZigClangSourceLocation,
+) !*ast.Node {
+    // expr != @bitCast(EnumType, @as(@TagType(EnumType), 0))
+
+    // @bitCast(Enum,
+    const bitcast = try transCreateNodeBuiltinFnCall(rp.c, "@bitCast");
+    const bitcast_enum_identifier = try generate_enum_node(rp, enum_ty, enum_source_loc);
+    try bitcast.params.push(bitcast_enum_identifier);
+    _ = try appendToken(rp.c, .Comma, ",");
+
+    // @as(
+    const cast_node = try transCreateNodeBuiltinFnCall(rp.c, "@as");
+
+    // @TagType(Enum),
+    const tag_type = try transCreateNodeBuiltinFnCall(rp.c, "@TagType");
+    const tag_type_enum_identifier = try generate_enum_node(rp, enum_ty, enum_source_loc);
+    try tag_type.params.push(tag_type_enum_identifier);
+    tag_type.rparen_token = try appendToken(rp.c, .RParen, ")");
+    try cast_node.params.push(&tag_type.base);
+    _ = try appendToken(rp.c, .Comma, ",");
+
+    // 0)
+    const zero = try transCreateNodeInt(rp.c, 0);
+    try cast_node.params.push(zero);
+    cast_node.rparen_token = try appendToken(rp.c, .RParen, ")");
+
+    try bitcast.params.push(&cast_node.base);
+    bitcast.rparen_token = try appendToken(rp.c, .RParen, ")");
+
+    // expr != @bitCast(EnumType, @as(@TagType(EnumType), 0))
+    return transCreateNodeNotEqual(rp, scope, expr, &bitcast.base);
+}
+
+fn transBoolExpr(
+    rp: RestorePoint,
+    scope: *Scope,
+    expr: *const ZigClangExpr,
+    used: ResultUsed,
+    lrvalue: LRValue,
+) !*ast.Node {
+    var res = try transExpr(rp, scope, expr, used, lrvalue);
+
+    switch (res.node.id) {
+        .InfixOp => switch (@ptrCast(*const ast.Node.InfixOp, &res.node).op) {
```

Comment:
Use `Node.cast`.

Reply:
Not sure I'm a fan of the fact that it can return null, might be better if it just does a compile time failure?

Or is that not possible in all scenarios? If so, maybe adding a separate `tryCast` would be a better option?

Reply:
You already checked that it is an InfixOp, it can't fail.

Reply:
If that's the case I might as well just do the cast, right?

Does the compile time check buy me anything? I guess it would give me a NULL segfault at runtime instead of reading random bits of the stack. But why does the cast bother making it a runtime failure, and not just compile error?

Reply:
`@ptrCast` is generally unsafe, and a blunt instrument that should be avoided when a more precise tool is available.

Indeed, your usage of `@ptrCast` here is incorrect, and is invoking undefined behavior. Zig does not guarantee order of struct fields; the `base` field of `InfixOp` is not guaranteed to be at offset 0. You'll notice that `Node.cast` uses `@fieldParentPtr`, not `@ptrCast`. The only reason your code runs in debug builds is because (1) you got (un)lucky and (2) #168 isn't implemented yet.

`Node.cast` *is* a "try cast", that's why it returns an optional pointer. `res.node.cast(ast.Node.InfixOp).?` does the cast, and asserts that the cast succeeded, returning a non-optional pointer.

> Does the compile time check buy me anything?

There is no compile time check, it's doing a runtime check to make sure the type is correct. When you already did such a check (the `switch`) then the correct thing to do is assert that the check will succeed (by using `.?`).

> I guess it would give me a NULL segfault at runtime instead of reading random bits of the stack, in the odd chance I made a mistake or a refactoring somehow caused it to be wrong. 

Not a segfault, but an explicit panic message saying [attempt to unwrap null](https://ziglang.org/documentation/master/#Attempt-to-Unwrap-Null).

> But why does the cast bother making it a runtime failure, and not just compile error?

`*Node` is a runtime-known pointer, with a runtime-known `id` field. Such a compile error is impossible.

Reply:
You might benefit with a structure more like this:

```zig
if (res.node.cast(ast.Node.InfixOp)) |infix_op| {
    // ...
} else if (res.node.cast(ast.Node.PrefixOp)) |prefix_op| {
    // ...
} else if (res.node.id == .BoolLiteral) {
    return res.node;
}
```

Reply:
When I read `base.id == comptime typeToId(T)`, in my rush to get to sleep, I thought the whole expression was comptime. Indeed it's a runtime check because Node.id can't be known at compile time.

The code this comment is on compiled successfully when it was `TransResult.node` (for some reason). However it fortunately doesn't compile now that `TransResult` was removed in a refactor.

I was going to solve this by swapping the code to use `@fieldParentPtr(res, "base", ast.Node.InfixOp)`. Is there a reason the if structure would be preferable?

Reply:
`@fieldParentPtr` would be fine. My suggestion was based on semantic convenience. `switch` with `@fieldParentPtr` and my if-else chain there would do the same thing semantically, it's just a question of what is more maintainable.